server/src/main/java/org/elasticsearch/tasks/TaskManager.java

```diff
@@ -607,4 +602,53 @@ void unregisterChildNode(DiscoveryNode node) {
         }
     }
 
+    /**
+     * Start tracking a cancellable task with its tcp channel, so if the channel gets closed we can get a set of
+     * pending tasks associated that channel and cancel them as these results won't be retrieved by the parent task.
+     *
+     * @return a releasable that should be called when this pending task is completed
+     */
+    public Releasable startTrackingCancellableChannelTask(TcpChannel channel, CancellableTask task) {
+        assert cancellableTasks.containsKey(task.getId()) : "task [" + task.getId() + "] is not registered yet";
+        final ChannelPendingTaskTracker tracker = channelPendingTaskTrackers.computeIfAbsent(channel, k -> new ChannelPendingTaskTracker());
+        tracker.addTask(task);
```

Comment:
What if the channel is closed (and the close listener executes completely) just before this line executes? I think that means we never cancel the task added in this line.

Could we have a test with some concurrency to pick this up?

Reply:
Good catch. I added a test and strengthen this logic in https://github.com/elastic/elasticsearch/pull/56620/commits/bdfa4f24a4fd99a819ada8db4580e84bbdd12a22.